File: src/utils/session.ts
import fs from "fs/promises";
import fsSync from "fs";
import path from "path";
import crypto from "crypto";
import os from "os";

const BASE_SESSIONS_DIR = path.join(os.homedir(), ".minicode", "sessions");
const EXT = ".context.jsonl";

interface SessionHeader {
  model: string;
  totalTokens: number;
  msgCount: number;
}

export interface SessionData {
  model: string;
  messages: Array<{ role: string; content: any }>;
  totalTokens: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface SessionInfo {
  name: string;
  updatedAt: string;
}

function logError(operation: string, error: unknown): void {
  if (process.env.DEBUG) {
    console.error(`[SessionManager] ${operation}:`, error);
  }
}

export class SessionManager {
  private readonly projectHash: string;
  private readonly sessionsDir: string;

  constructor() {
    this.projectHash = this.computeProjectHash();
    this.sessionsDir = path.join(BASE_SESSIONS_DIR, this.projectHash);
  }

  private computeProjectHash(): string {
    const cwd = process.cwd();
    return crypto.createHash("md5").update(cwd).digest("hex").substring(0, 12);
  }

  private async ensureDir(): Promise<void> {
    await fs.mkdir(this.sessionsDir, { recursive: true });
  }

  private sessionPath(name: string): string {
    return path.join(this.sessionsDir, `${name}${EXT}`);
  }

  getProjectHash(): string {
    return this.projectHash;
  }

  getSessionDir(): string {
    return this.sessionsDir;
  }

  async list(): Promise<SessionInfo[]> {
    await this.ensureDir();
    const entries = await fs.readdir(this.sessionsDir).catch(() => []);
    const sessions: SessionInfo[] = [];

    for (const entry of entries) {
      if (!entry.endsWith(EXT)) continue;
      const name = entry.replace(EXT, "");
      try {
        const stat = await fs.stat(this.sessionPath(name));
        sessions.push({ name, updatedAt: stat.mtime.toISOString() });
      } catch {
        // skip unreadable files
      }
    }

    return sessions.sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
  }

  async listNames(): Promise<string[]> {
    await this.ensureDir();
    const entries = await fs.readdir(this.sessionsDir).catch(() => []);
    return entries.filter((e) => e.endsWith(EXT)).map((e) => e.replace(EXT, ""));
  }

  async getMostRecent(): Promise<string | null> {
    const sessions = await this.list();
    return sessions.length > 0 ? sessions[0].name : null;
  }

  async get(name: string): Promise<SessionData | null> {
    await this.ensureDir();
    const filePath = this.sessionPath(name);

    // Try JSONL format first
    try {
      const content = await fs.readFile(filePath, "utf-8");
      const lines = content.split("\n").filter((l) => l.trim());
      if (lines.length === 0) return null;

      const header: SessionHeader = JSON.parse(lines[0]);
      const messages: SessionData["messages"] = [];
      for (let i = 1; i < lines.length; i++) {
        try {
          messages.push(JSON.parse(lines[i]));
        } catch {
          // skip malformed lines
        }
      }

      return {
        model: header.model,
        messages,
        totalTokens: header.totalTokens,
      };
    } catch {
      // JSONL not found, try legacy .json format
    }

    // Fallback: legacy .json format
    const legacyPath = path.join(this.sessionsDir, `${name}.json`);
    try {
      const content = await fs.readFile(legacyPath, "utf-8");
      return JSON.parse(content);
    } catch (error) {
      logError(`failed to read session ${name}`, error);
      return null;
    }
  }

  async save(name: string, data: SessionData): Promise<void> {
    await this.ensureDir();
    const filePath = this.sessionPath(name);
    const tmpPath = filePath + ".tmp";

    const header: SessionHeader = {
      model: data.model,
      totalTokens: data.totalTokens,
      msgCount: data.messages.length,
    };

    const lines: string[] = [JSON.stringify(header)];
    for (const msg of data.messages) {
      lines.push(JSON.stringify(msg));
    }

    // Atomic write: write to tmp then rename
    await fs.writeFile(tmpPath, lines.join("\n") + "\n");
    await fs.rename(tmpPath, filePath);
  }

  async delete(name: string): Promise<void> {
    await this.ensureDir();
    // Try JSONL first, then legacy .json
    const paths = [this.sessionPath(name), path.join(this.sessionsDir, `${name}.json`)];
    for (const filePath of paths) {
      try {
        await fs.unlink(filePath);
      } catch {
        // ignore
      }
    }
  }

  async rename(oldName: string, newName: string): Promise<void> {
    await this.ensureDir();
    const oldPath = this.sessionPath(oldName);
    const newPath = this.sessionPath(newName);
    try {
      await fs.rename(oldPath, newPath);
    } catch (error) {
      logError(`failed to rename session ${oldName} to ${newName}`, error);
    }
  }
}

export const sessionManager = new SessionManager();

File: internal/session.go
package internal

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

// SessionData contains the serialisable session state.
type SessionData struct {
	Model       string `json:"model"`
	Messages    []any  `json:"messages"`
	TotalTokens int    `json:"totalTokens"`
	CreatedAt   string `json:"createdAt"`
	UpdatedAt   string `json:"updatedAt"`
}

// SessionInfo is summary metadata for a saved session.
type SessionInfo struct {
	Name      string
	UpdatedAt string
}

// SessionManager persists sessions to ~/.minicode/sessions/<projectHash>/.
type SessionManager struct {
	sessionsDir string
}

// NewSessionManager creates a session manager for the current working directory.
func NewSessionManager() *SessionManager {
	cwd, _ := os.Getwd()
	hash := fmt.Sprintf("%x", md5.Sum([]byte(cwd)))[:12]
	home, _ := os.UserHomeDir()
	return &SessionManager{
		sessionsDir: filepath.Join(home, ".minicode", "sessions", hash),
	}
}

func (m *SessionManager) ensureDir() error { return os.MkdirAll(m.sessionsDir, 0o755) }

// List returns all saved sessions sorted by update time (newest first).
func (m *SessionManager) List() ([]SessionInfo, error) {
	_ = m.ensureDir()
	entries, err := os.ReadDir(m.sessionsDir)
	if err != nil {
		return nil, nil
	}
	var sessions []SessionInfo
	for _, e := range entries {
		if filepath.Ext(e.Name()) != ".json" {
			continue
		}
		name := e.Name()[:len(e.Name())-5]
		data, err := m.Get(name)
		if err == nil && data != nil {
			sessions = append(sessions, SessionInfo{Name: name, UpdatedAt: data.UpdatedAt})
		}
	}
	sort.Slice(sessions, func(i, j int) bool { return sessions[i].UpdatedAt > sessions[j].UpdatedAt })
	return sessions, nil
}

// MostRecent returns the name of the most recently updated session.
func (m *SessionManager) MostRecent() (string, error) {
	sessions, err := m.List()
	if err != nil || len(sessions) == 0 {
		return "", err
	}
	return sessions[0].Name, nil
}

// Get loads a named session.
func (m *SessionManager) Get(name string) (*SessionData, error) {
	_ = m.ensureDir()
	b, err := os.ReadFile(filepath.Join(m.sessionsDir, name+".json"))
	if err != nil {
		return nil, err
	}
	var data SessionData
	if err := json.Unmarshal(b, &data); err != nil {
		return nil, err
	}
	return &data, nil
}

// Save persists session data.
func (m *SessionManager) Save(name string, data *SessionData) error {
	_ = m.ensureDir()
	now := time.Now().Format(time.RFC3339)
	data.UpdatedAt = now
	if data.CreatedAt == "" {
		data.CreatedAt = now
	}
	b, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(m.sessionsDir, name+".json"), b, 0o644)
}

// Delete removes a saved session.
func (m *SessionManager) Delete(name string) error {
	_ = m.ensureDir()
	return os.Remove(filepath.Join(m.sessionsDir, name+".json"))
}

// Rename renames a session file.
func (m *SessionManager) Rename(oldName, newName string) error {
	_ = m.ensureDir()
	return os.Rename(
		filepath.Join(m.sessionsDir, oldName+".json"),
		filepath.Join(m.sessionsDir, newName+".json"),
	)
}
